src/main/kotlin/com/xtyrrell/maxonbank/query/BankInfoProjector.kt

```diff
@@ -0,0 +1,26 @@
+package com.xtyrrell.maxonbank.query
+
+import com.xtyrrell.maxonbank.coreapi.*
+import org.axonframework.eventhandling.EventHandler
+import org.axonframework.queryhandling.QueryHandler
+import org.springframework.stereotype.Component
+
+@Component
+class BankInfoProjector {
```

Comment:
As far as I understand, Axon should always replay the event log on startup. Otherwise it would be only viable to store state in the DB, but I think Axon considers storing state in the DB the alternative, and event-sourcing state by maintaining variables in an @ Aggregate class the typical route.

> In the Aggregate main page we have seen how to create an Aggregate backed by Event Sourcing. In other words, the storage method for an Event Sourced Aggregate is by replaying the events which constitute the changes on the Aggregate.
> An Aggregate can however be stored as-is too. When doing so, the Repository used to save and load the Aggregate, is the GenericJpaRepository. The structure of a state-stored Aggregate is a little different from an Event Sourced Aggregate:

from https://docs.axoniq.io/reference-guide/axon-framework/axon-framework-commands/modeling/state-stored-aggregates

It's also done in the Axon course I did.

So from the looks of it I think it should be safe, but would be good to do more thorough research.

Reply:
I think it's super clean to seperate out event listeners into different classes based on their responsibilitie -- what do you think about having `BankInfoProjector` seperate from `AccountProjector`?

Reply:
I think an event can only belong to one aggregate, and an aggregate only "hears" the events that have the same aggregateIdentifier. So you won't be able to have an AccountAggregate (that EventSources the account level balance), and an BankAggragate (that listens to the events from AccountAggregates and maintains an overall balance).

You should be able to setup event listeners, that tracks how far it's listened, and only listens to new events, and then just ensure that everytime your component (BankProjector) goes into memory it starts tracking from the start again. https://docs.axoniq.io/reference-guide/axon-framework/events/event-processors

I haven't really tried to do this, but I like it as a concept, going to do some experimenting. You would probably want to snapshot though, so you don't have to replay every event for every account every time your app restarts.

Keeping the Projectors separate is definitely a great idea, it lets you build fit for purpose queries. It's another thing that feels "wrong" when coming from a traditional non CQRS background, because you often end up duplicating data (maybe saving a ledger entry in AccountProjector and BankInfoProjector), but the info you save in each one is based on the audience that is going to consume it, like maybe in BankInfoProjector you save some debugging info for internal use like what service initiated this transfer, but in AccountProjector you don't need that because an end user would never see it.

Reply:
Great, thanks for the feedback :)